tests/virt_autotest/esxi_open_vm_tools.pm

```diff
@@ -0,0 +1,209 @@
+}
+
+sub wait_for_vm_bootup {
+    my $vm_ip       = shift;         # vm ip address
+    my $if_pingable = shift;         # 1 or 0
+    my $times       = shift // 6;    # the number of ping execution
+    my $count       = shift // 3;    # ping waits for the count
+    my $interval    = shift // 5;    # wait secs between each ping
+    my $cmd;
+
+    if ($if_pingable) {
+        $cmd = "ping -c$count -i$interval $vm_ip";
+    } else {
+        $cmd = "! ping -c$count -i$interval $vm_ip";
+    }
+
+    while ($times && console('svirt')->run_cmd($cmd)) {
+        $times--;
+    }
+}
+
+sub login_vm_console {
+    my $vm_name = shift;
+
+    reset_consoles;
+    console('svirt')->start_serial_grab($vm_name);
+    select_console('sut');
+    assert_screen('grub2', 90);
+    send_key 'ret';
+    select_console('root-console', timeout => 180);
+}
+
+sub set_random_time {
+    my $n_hrs     = (sprintf "%02d", rand(24));
+    my $n_min     = (sprintf "%02d", rand(59));
+    my $n_sec     = (sprintf "%02d", rand(59));
+    my $rand_time = script_output("date -u -s $n_hrs:$n_min:$n_sec +'\%F \%T'");
+
+    return $rand_time;
+}
+
+sub get_epoch_time {
+    my ($datetime, $epoch, $sec, $min, $hour, $mday, $mon, $year);
+    $datetime = $_[0];
+
+    my @datetime = reverse(split /-|:|\s|\//, $datetime);
+    foreach (@datetime) {
+        s/^0//;
+        $epoch .= $_ . ",";
+    }
+    $epoch =~ s/,$//;
+    ($sec, $min, $hour, $mday, $mon, $year) = split /,/, $epoch;
+    $year -= 1900;
+    $mon  -= 1;
+    $epoch = timelocal($sec, $min, $hour, $mday, $mon, $year);
+
+    return $epoch;
+}
+
+sub do_power_management_tests {
+    my ($vm_id, $vm_ip) = @_;
+
+    record_info('Power Manangement Tests');
+    select_console('svirt');
+
+    record_info('Guest Power Off');
+    check_vm_power_state($vm_id, $vm_ip, 'power.off', 0, 1, $VM_POWERED_ON, $VM_POWERED_OFF);
+
+    record_info('Guest Power On');
+    check_vm_power_state($vm_id, $vm_ip, 'power.on', 1, 1, $VM_POWERED_OFF, $VM_POWERED_ON);
+
+    record_info('Guest Power Reset');
+    check_vm_power_state($vm_id, $vm_ip, 'power.reset', 1, 1, $VM_POWERED_ON, $VM_POWERED_ON);
+
+    record_info('Guest Power Restart');
+    check_vm_power_state($vm_id, $vm_ip, 'power.restart', 1, 1, $VM_POWERED_ON, $VM_POWERED_ON);
+
+    record_info('Guest Power Shutdown');
+    check_vm_power_state($vm_id, $vm_ip, 'power.shutdown', 0, 1, $VM_POWERED_ON, $VM_POWERED_OFF);
+}
+
+sub do_networking_tests {
+    my ($vm_name, $vm_id, $vm_ip) = @_;
+
+    record_info('Networking Tests');
+    check_vm_power_state($vm_id, $vm_ip, 'power.on', 1, 0, $VM_POWERED_OFF, $VM_POWERED_ON);
+
+    my $vswitch_name = esxi_vm_network_binding($vm_id);
+    record_info('VSwitch Name', $vswitch_name);
+
+    login_vm_console($vm_name);
+
+    assert_script_run('ping -4 -c3 openqa.suse.de');
+    assert_script_run('ping -4 -c3 www.suse.com');
```

Comment:
What about "ping -I $vm_ip -4 -c3 xxxx" ? So your ping bind to your preferred ipv4 address ?

Reply:
It's a good suggestion. Agree on this. Updated.